deltachat-ios/AppDelegate.swift

```diff
@@ -386,25 +389,41 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterD
 
         DispatchQueue.main.asyncAfter(deadline: .now() + 10) { [weak self] in
             logger.info("⬅️ finishing fetch")
+            self?.endFetch(completionHandler: completionHandler)
+        }
+    }
+
+    private func endFetch(completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
+        if fetchBackgroundTaskEnding {
+            logger.info("⬅️ fetch already ending")
+            return
+        }
+        fetchBackgroundTaskEnding = true
+
+        if !self.appIsInForeground() {
+            self.dcContext.stopIo()
+        }
 
+        // stopIo() may result in events spanning their own backgroundTask.
+        // a gap between these tasks and this one may result in 0xdead10cc exceptions;
+        // therefore, we wait a little moment.
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.8) { [weak self] in
             guard let self = self else {
+                logger.info("⬅️ fetch done, self lost")
                 completionHandler(.failed)
                 return
             }
-            if !self.appIsInForeground() {
-                self.dcContext.stopIo()
-            }
 
-            // to avoid 0xdead10cc exceptions, scheduled jobs need to be done before we get suspended;
-            // we increase the probabilty that this happens by waiting a moment before calling completionHandler()
-            DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
-                logger.info("⬅️ fetch done")
-                completionHandler(.newData)
+            logger.info("⬅️ fetch done")
+            completionHandler(.newData)
+            self.unregisterFetchBackgroundTask()
```

Comment:
maybe we should move that line before the completion handler is executed in order to be sure this is called before the suspension of the app happened.

Reply:
i think, this is fine, the backgroundTask avoids the suspension.